File: java/316_RemoveDuplicateLetters.java
class Solution {
    public String removeDuplicateLetters(String s) {
        HashMap<Character, Integer> letterCounts = new HashMap<Character, Integer>();
        for (int i = 0; i < s.length(); ++i) letterCounts.put(s.charAt(i), letterCounts.getOrDefault(s.charAt(i), 0) + 1);
        
        List<Character> letterStack = new ArrayList<Character>();
        HashSet<Character> isInStack = new HashSet<Character>();
        for (int i = 0; i < s.length(); ++i) {
            Character c = s.charAt(i);
            if (!isInStack.contains(c)) {
                while (!letterStack.isEmpty() && c < letterStack.get(letterStack.size() - 1) && letterCounts.get(letterStack.get(letterStack.size() - 1)) > 0) {
                    isInStack.remove(letterStack.get(letterStack.size() - 1));
                    letterStack.remove(letterStack.size() - 1);
                }
                letterStack.add(c);
                isInStack.add(c);
            }
            letterCounts.put(c, letterCounts.get(c) - 1);
        }
        StringBuilder sb = new StringBuilder();
        for (Character c : letterStack) sb.append(c);
        return sb.toString();
    }
}

File: swift/316_RemoveDuplicateLetters.swift
class Solution {
    func removeDuplicateLetters(_ s: String) -> String {
        var letterCounts = [Character:Int]()
        for c in s {
            letterCounts[c] = 1 + (letterCounts[c] ?? 0)
        }
        
        var stack = [Character]()
        var stackSet = Set<Character>()
        for c in s {
            if !stackSet.contains(c) {
                while !stack.isEmpty && c < stack.last! && letterCounts[stack.last!]! > 0 {
                    stackSet.remove(stack.removeLast())
                } 
                stack.append(c)
                stackSet.insert(c)
            }
            letterCounts[c] = (letterCounts[c] ?? 0) - 1
        }
        
        var result = ""
        for c in stack { result.append(c) }
        return result
    }
}
